clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -237,6 +245,66 @@ class BuiltinNameEmitter {
   // same entry (<I1, I2, I3>).
   MapVector<BuiltinIndexListTy *, BuiltinTableEntries> SignatureListMap;
 };
+
+/// Generate a json file representing the builtins as understood by clang.
+/// This allows the generation of a library interface.
+class JSONBuiltinInterfaceEmitter : public BuiltinNameEmitter {
+  struct TypeDesc {
+    TypeDesc(const Record *T);
+    // Return a string representing the "base" type (ignores pointers).
+    std::string GetBaseTypeAsStr() const;
+
+    enum UnderlyingType { BOOL, INT, UINT, FLOAT, BUILTIN };
+
+    llvm::StringRef Name;
+    // Size of the vector (if applicable).
+    int VecWidth;
+    // Size of the element in bits.
+    int ElementSize;
+    // The underlying type represented by the record.
+    UnderlyingType ElementType;
+    // Is a pointer.
+    bool IsPointer;
+    // "const" qualifier.
+    bool IsConst;
+    // "volatile" qualifier.
+    bool IsVolatile;
+    std::string AddrSpace;
```

Comment:
Since these are always being initialized with string literals, this could be a StringRef instead!

Reply:
done